feat: add types to openapi

astagi opened the pull request:
I removed `tests/test_api_oas3.py` because it uses api module, which depends by `doc` module

This PR allows to:

- expose `openapi`
- use types from `openapi`

You can document your OAS3 API in this way

```python
from sanic import Sanic
from sanic.response import json as json_response
from sanic_openapi import openapi, oas3_blueprint


app = Sanic("AwesomeApi")
app.blueprint(oas3_blueprint)


class Car:
    manufacturer = openapi.String(description="Car manufacturer", required=True)
    model = openapi.String(description="Car model", required=True)
    production_date = openapi.Date(description="Car year", required=True)


class Garage:
    spaces = openapi.Integer(description="Space available in the garage", required=True)
    cars = openapi.Array(Car, required=True)


@app.post('/garage')
@openapi.description("Create a new garage")
@openapi.body(
    { "application/json" : Garage },
    description="Body description",
    location="/garage",
    required=True,
)
@openapi.response(
    201, { "application/json" : Garage }, "A new Garage is created"
)
async def create_garage(request):
    return json_response(request.json, 201)

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8000, debug=True)
```

astagi commented:
@ahopkins @artcg

I'll add more tests later.

artcg commented:
Nice to see a usage example like that, we should get that code snippet in /examples maybe! Looks good, will leave a comment

artcg commented on sanic_openapi/oas3/types.py:
```diff
@@ -185,7 +184,7 @@ class Array(Schema):
     uniqueItems: False
 
     def __init__(self, items: Schema, **kwargs):
```
typing for 'items' can be set to 'Any' or removed now, nice changes :)

astagi replied:
You're right @artcg , done!

astagi commented:
@artcg I've just added an example for OAS3. Everything works fine, except header parameters. I added AUTHORIZATION header parameter in `car_put`

```python
@openapi.parameter("AUTHORIZATION", str, location="header")
@openapi.response(200, {"application/json" : Car})
def car_put(request, car_id):
    return json(test_car)
```

Swagger interface detects the header parameter correctly but when I try out the API call it's not included in the curl command. Any ideas?

artcg commented:
Looks great, taking a look at that header issue now

artcg commented:
Ah in openapi3: 

header parameters for "authorization" are ignored as it recommends to use securityschema instead, so can replace that header with something else for the example maybe...

If in is "header" and the name field is "Accept", "Content-Type" or "Authorization", the parameter definition SHALL be ignored.

astagi commented:
hey @artcg thanks for having looked at this! I'll replace it with `SecurityScheme` then and see if it works as expected!

astagi commented:
@artcg I updated SecurityScheme to make it working in cars example. I don't know if it's my fault, but itooks like Swagger interface is not working as expected, in `put car` method when `@openapi.secured` decorator is applied I see a little padlock

![image](https://user-images.githubusercontent.com/537363/111953033-551ce400-8ae6-11eb-8815-e7060a8c6c15.png)

every time I try to call the API, no value is included in the header by curl again. 

Specs also suggest to define `securitySchemes` object into `components`, but ComponentsBuilder  is commented inside `builders.py`. Any ideas?

artcg commented:
Ah looks like we will need the Components field supported for this which will be a bit of a chore, will push something later in the week, until then, lets merge what looks good so far, can you open a new PR from a new branch based upon this one for the security changes?

artcg reviewed (approved):
Im going to merge, great job so far! I would say we dont need to get SecuritySchemas working for the 0.6.3 release next week, so just raised a NotImplementedError there as it would be nice to take a break also :)

astagi commented:
eheh totally agree @artcg! I'm going to start working on oas3 naming changes and folders structure.. Thank you for your support!
